lib/filterx/func-sdata.c

```diff
@@ -0,0 +1,297 @@
+}
+
+static void
+_free(FilterXExpr *s)
+{
+  FilterXFunctionIsSdataFromEnteprise *self = (FilterXFunctionIsSdataFromEnteprise *) s;
+
+  g_free(self->str_literal);
+  filterx_function_free_method(&self->super);
+}
+
+
+FilterXExpr *
+filterx_function_is_sdata_from_enterprise_new(const gchar *function_name, FilterXFunctionArgs *args, GError **error)
+{
+  FilterXFunctionIsSdataFromEnteprise *self = g_new0(FilterXFunctionIsSdataFromEnteprise, 1);
+  filterx_function_init_instance(&self->super, function_name);
+
+  if (!_extract_args(self, args, error))
+    goto error;
+  self->super.super.eval = _eval;
+  self->super.super.free_fn = _free;
+  filterx_function_args_free(args);
+  return &self->super.super;
+
+error:
+  filterx_function_args_free(args);
+  filterx_expr_unref(&self->super.super);
+  return NULL;
+}
+
+
+FilterXObject *
+filterx_simple_function_has_sdata(FilterXExpr *s, GPtrArray *args)
+{
+  if (args && args->len != 0)
+    {
+      filterx_simple_function_argument_error(s, "Incorrect number of arguments", FALSE);
+      return NULL;
+    }
+
+  FilterXEvalContext *context = filterx_eval_get_context();
+  LogMessage *msg = context->msgs[0];
+  return filterx_boolean_new(msg->num_sdata != 0);
+}
+
+static gboolean
+_should_create_new_dict(const gchar *previous_sd_id, const gchar *current_sd_id)
+{
+  return strcmp(previous_sd_id, current_sd_id) != 0;
+}
+
+static gboolean
+_extract_sd_param_name(const gchar *key, gssize key_len, GString *dest_buf)
+{
+  /*skip '.SDATA.'*/
+  if (key_len < SDATA_PREFIX_LEN)
+    return FALSE;
+
+  gchar *last_dot = strrchr(key, '.');
+  if (!last_dot)
+    return FALSE;
+
+  const gchar *sd_param_name_start = last_dot + 1;
+  g_string_truncate(dest_buf, 0);
+  g_string_assign(dest_buf, sd_param_name_start);
+  return TRUE;
+}
+
+static gboolean
+_extract_sd_id(const gchar *key, gssize key_len, GString *dest_buf)
+{
+  /*skip '.SDATA.'*/
+  if (key_len < SDATA_PREFIX_LEN)
+    return FALSE;
+
+  gchar *last_dot = strrchr(key, '.');
+  if (!last_dot)
+    return FALSE;
+
+  const gchar *sd_id_start = key + SDATA_PREFIX_LEN;
+  g_string_truncate(dest_buf, 0);
+  g_string_append_len(dest_buf, sd_id_start, last_dot - sd_id_start);
+  return TRUE;
+}
+
+static gboolean
+_insert_into_dict(FilterXObject *dict, const gchar *key, const gchar *value)
+{
+  FilterXObject *fob_key = filterx_string_new(key, -1);
+  FilterXObject *fob_value = filterx_string_new(value, -1);
+  gboolean res = filterx_object_set_subscript(dict, fob_key, &fob_value);
+
+  filterx_object_unref(fob_key);
+  filterx_object_unref(fob_value);
+
+  return res;
+}
+
+static gboolean
+_generate(FilterXExprGenerator *s, FilterXObject *fillable)
+{
+  FilterXEvalContext *context = filterx_eval_get_context();
+  LogMessage *msg = context->msgs[0];
+
+  ScratchBuffersMarker marker;
+  GString *current_sd_id = scratch_buffers_alloc_and_mark(&marker);
+  GString *next_sd_id = scratch_buffers_alloc();
+  GString *param_name = scratch_buffers_alloc();
```

Comment:
Copying to `GString`s is slow and unnecessary. The `sd_id` and `sd_param` strings could be `const gchar *`s with `gsize` len, kind of "Slices" in the original string.